### PR TITLE
improve documentation of the Widom insertion method

### PR DESCRIPTION
**Pull request by jonaslandsgesell**
This PR adds documentation for the Widom insertion method.

In addition it adds the feature to change reaction constants and delete reactions. This is functionality which Oleg (@helvrud) suggested and which gives some more flexibility to the user.

### CODE REVIEW

**Comment by RudolfWeeber** [comment]
@kosovan, would you be willing to review this?

**Comment by kosovan** [comment]
> @kosovan, would you be willing to review this?

@RudolfWeeber I should be able to do it on Tuesday or Wednesday. I am not sure if that is not too late but I have some teaching duties that cannot be postponed.

**Comment by jonaslandsgesell** [comment]
@helvrud I implemented the ability to change the reaction constant or delete reactions you already added:

deletion of a given reaction:
```
RE.delete_reaction(1)
```
change of the reaction constant (back and forward reaction simultaneously):

```
reaction_id=0
RE.change_reaction_constant(reaction_id, new_reaction_constant)
```

File: doc/sphinx/advanced_methods.rst
Comment (kosovan) [inline doc/sphinx/advanced_methods.rst:1858]: I find this section title rather confusing: the section title suggests that one can use the reaction ensemble to simulate a constant pH. What it describes below it the Constant-pH ensemble, implemented as a method inside the ReactionEnsemble class of Espresso. Please, here and below strictly distinguish between the `ReactionEnsemble` as a name for the respective class in Espresso, and the names of the actual thermodynamic ensembles: "reaction ensemble" a "constant-pH ensemble".

File: doc/sphinx/advanced_methods.rst
Comment (kosovan) [inline doc/sphinx/advanced_methods.rst]: Please, explain what is meant by "0th registered reaction".

File: doc/sphinx/advanced_methods.rst
Comment (kosovan) [inline doc/sphinx/advanced_methods.rst]: This equation won't fit in one line. Please, add a line break at a suitable position.

File: doc/sphinx/advanced_methods.rst
Comment (kosovan) [inline doc/sphinx/advanced_methods.rst]: This equation won't fit on a single line. Please, add line break(s) at suitable position(s).

File: doc/sphinx/advanced_methods.rst
Comment (kosovan) [reply doc/sphinx/advanced_methods.rst]: `\mathrm{\emptyset \to type_B` fails to render in my browser (close-bracket missing?)

File: doc/sphinx/advanced_methods.rst
Comment (kosovan) [inline doc/sphinx/advanced_methods.rst]: There is no next insertion mentioned above. I propose to re-phrase: "If another insertion (reaction) is defined, then the chemical potential of the particle inserted in that reaction can be measured via calling ..."

File: doc/sphinx/advanced_methods.rst
Comment (kosovan) [inline doc/sphinx/advanced_methods.rst]: I do not understand what is meant by the "volume average" and how it differs from the "ensemble average". To my understanding, the "ensemble average" implies averaging over different positions within the given volume.

File: doc/sphinx/advanced_methods.rst
Comment (kosovan) [inline doc/sphinx/advanced_methods.rst:1991]: Please, describe under which conditions one should desire such behaviour, and under which conditions such behaviour should be undesired. To my understanding, the Widom method never really inserts a new particle, it only attemts to do so and computes the associated free energy change. So, such situation should never occur.

File: doc/sphinx/advanced_methods.rst
Comment (kosovan) [reply doc/sphinx/advanced_methods.rst:1858]: Perhaps this is a more general issue: the `reaction_ensemble` class now contains several methods, not just the eponymic `ReactionEnsemble`. Perhaps it would be more appropriate to give a more generic name to the whole class, e.g. `reaction_mehods` or `chemical_reactions`.

File: src/python/espressomd/reaction_ensemble.pyx
Comment (kosovan) [inline src/python/espressomd/reaction_ensemble.pyx:776]: This assumption is almost never fulfilled, so the provided error estimates are useless in most cases. What is worse, the output gives no chance to the user to check how well the assumption was fulfilled. I think that it would be more useful to output the whole time-series, and let the user estimate the error using his own favorite method.

File: src/python/espressomd/reaction_ensemble.pyx
Comment (kosovan) [inline src/python/espressomd/reaction_ensemble.pyx]: It it not clear in the given context what it means to "register the insertion move". I propose to rephrase this as: "The reaction with given reaction_id has to be defined before."

File: src/python/espressomd/reaction_ensemble.pyx
Comment (kosovan) [inline src/python/espressomd/reaction_ensemble.pyx:779]: I think that the "inverse widom scheme" is not documented anywhere.

File: doc/sphinx/advanced_methods.rst
Comment (kosovan) [inline doc/sphinx/advanced_methods.rst]: ... due to the insertion of a new particle, or a group of particles

**Review by kosovan — CHANGES_REQUESTED** [review]

File: src/python/espressomd/reaction_ensemble.pyx
Comment (jonaslandsgesell) [reply src/python/espressomd/reaction_ensemble.pyx:779]: The inverse widom scheme refers to a scheme like this one: http://www.mtmml.gr/library/downloads/docs/documents/publications/paprer3.pdf
How to use it was documented before but I removed it after talking to Oleg... also since special adaptions to the formulas for calculating the excess chemical potential have to be made in the core for it to be fully correct. Also it is untested. The comments are there, if somebody wanted to reintroduce this scheme. Giving more details here goes beyond this PR.

File: src/python/espressomd/reaction_ensemble.pyx
Comment (jonaslandsgesell) [reply src/python/espressomd/reaction_ensemble.pyx:776]: This goes beyond documentation. Can you open a PR which fixes the corresponding issue #3283 ?

File: doc/sphinx/advanced_methods.rst
Comment (jonaslandsgesell) [reply doc/sphinx/advanced_methods.rst:1858]: True, I opened an issue for this #3284

File: doc/sphinx/advanced_methods.rst
Comment (jonaslandsgesell) [reply doc/sphinx/advanced_methods.rst]: a volume average is caluclated via f_V(t):=\frac{1}{V}int_V f(\vec{r},t) d^3r. This is spacial averaging of a quantity f.
The ensemble and volume average average of that quantity f could on the other hand be obtained via the time average (if the system is ergodic) of the volume average: <f_V>=1/T \int_0^T f_V(t) dt

File: doc/sphinx/advanced_methods.rst
Comment (jonaslandsgesell) [reply doc/sphinx/advanced_methods.rst:1991]: Well, one could be interested in the excess chemical potential change for inserting one particle of type B, one particle of type C while removing one particle of type A.
i.e. the excess chemical potential associated with A-> B+C.
Now B would always be placed at the position of the former A. It is not clear whether this "correlation" is intended by the user. 

PS: In principle the deletion scheme is valid for measuring the excess chemical potential of a particle... So this is untested but could be done in principle. The warning above is for those who want to push the limits and maybe do something new

File: doc/sphinx/advanced_methods.rst
Comment (jonaslandsgesell) [reply doc/sphinx/advanced_methods.rst]: `}` is missing

File: src/python/espressomd/reaction_ensemble.pyx
Comment (jonaslandsgesell) [reply src/python/espressomd/reaction_ensemble.pyx]: You define a certain insertion move (e.g. emptyset -> 1 A, for inserting one particle of type A, or emptyset -> 1 A+1 B for inserting one particle of type A and one particle of type B, via calling add_reaction() previously). That is why there is a reference to add_reaction.

File: src/python/espressomd/reaction_ensemble.pyx
Comment (kosovan) [reply src/python/espressomd/reaction_ensemble.pyx:779]: @jonaslandsgesell I thought that there was a consensus that undocumented code and features should not be present in the master.

File: src/python/espressomd/reaction_ensemble.pyx
Comment (kosovan) [reply src/python/espressomd/reaction_ensemble.pyx]: As mentioned before
> It is not clear in the given context.

Please, explain it to the reader, not to me.

One should keep the terminology consistent. If the documentation of the reaction ensemble uses the term _define_ a reaction and not _register_, then you should use the same term here.

File: doc/sphinx/advanced_methods.rst
Comment (kosovan) [reply doc/sphinx/advanced_methods.rst]: I disagree. The _ensemble average_ is by definition an average over all possible configurations in a given ensemble. 
Only if a given quantity can be defined locally, and if the system is sufficiently big, then the ensemble average can be approximated by averaging over different volume elements.

Only if the system is ergodic, then it can be approximated by time-averaging.

The term _volume average_ is not established in statistical mechanics (to my knowledge) and should not be used, or should be clearly defined before it is used. Otherwise, it does not explain anything to the reader.

File: src/python/espressomd/reaction_ensemble.pyx
Comment (kosovan) [reply src/python/espressomd/reaction_ensemble.pyx:776]: I am afraid that I cannot open such PR within next 10 days because I am busy with other staff.

File: doc/sphinx/advanced_methods.rst
Comment (jonaslandsgesell) [reply doc/sphinx/advanced_methods.rst]: Please have a look at the above formula (see sphinx doc). It involves averages "over the position of particles N_1+1 and N_2+1" (this is how Frenkel calls these "volume averages", below eq. 7.2.5 in his book)... I do not mind calling them "averages over the position of particle N_1+1 and N_2+1", which is what i call a volume average.

Comment (kosovan) [reply doc/sphinx/advanced_methods.rst]: I think that _averages over particle positions_ is OK. What I see as a problem is the other term _ensemble average_ because it is ambiguous. If you mean averaging over time, then please use the term _time average_

Comment (jonaslandsgesell) [reply doc/sphinx/advanced_methods.rst]: for certain systems there is no concept of time (e.g. when propagating the system with MC moves only). 
Frenkel & Smit call then involved average "<...>_N, ... [the] canonical ensemble average, averaging over the configuration space of the N-particle system".

Comment (kosovan) [reply doc/sphinx/advanced_methods.rst]: Some authors use the term "MC time". If you do not want to use the term _time average_, then we should think of a different term. Because _ensemble average_ is a general term that includes all kinds of averaging that one might perform within an ensemble.

File: src/python/espressomd/reaction_ensemble.pyx
Comment (jonaslandsgesell) [reply src/python/espressomd/reaction_ensemble.pyx:776]: This is part of another issue, resolving it here.

File: doc/sphinx/advanced_methods.rst
Comment (jonaslandsgesell) [reply doc/sphinx/advanced_methods.rst:1858]: This is part of another issue, resolving it here.

File: src/python/espressomd/reaction_ensemble.pyx
Comment (jonaslandsgesell) [reply src/python/espressomd/reaction_ensemble.pyx:779]: There is no undocumented code here. This is a check for allowed reaction_ids. The comment explains to some degree why the check involves the number 2: insertions and deletions are always defined in a pariwise manner:
calling add_reaction(product_types=[type_A],...)
defines the insertion and the deletion move... so you have reaction_ids 0 and 1 defined in the core. For the python interface the deletions are not allowed to be accessed via the code above...

**Comment by RudolfWeeber** [comment]
@jonaslandsgesell this pr adds functionallity in addition to docs. Please edit the description to explain this PRś intention

**Review by RudolfWeeber — APPROVED** [review]

**Review by RudolfWeeber — APPROVED** [review]

**Comment by jonaslandsgesell** [comment]
@kosovan could you mark your comments as resolved? Otherwise the merge cannot proceed and people do not find documentation about the method. If there are remaining problems please open a new issue for them.

**Comment by KaiSzuttor** [comment]
@jonaslandsgesell that's an issue of bors actually

**Comment by jngrad** [comment]
@KaiSzuttor it's actually a feature, `bors` won't merge if anyone requested a change, until that person approves the PR. You can dismiss kosovan's review if you want to proceed with `bors`.

**Comment by jngrad** [comment]
Should this PR be milestoned as 4.2 or 4.1.2?

**Comment by KaiSzuttor** [comment]
@jngrad personally I think only github should decide whether the PR is ready to merge. Also, it seems that I'm not able to dismiss the review of @kosovan. So we can only manually merge.

**Comment by KaiSzuttor** [comment]
image

... and the API says it's ready to merge

**Comment by RudolfWeeber** [comment]
>Should this PR be milestoned as 4.2 or 4.1.2?

Ideally, this should have been two PRs. Since the potential of side effects is small in this case, I’d be OK with 4.1.2

**Comment by kosovan** [comment]
Hey guys,

I apologize for the late response. I was not aware that this was waiting for me to mark my requests as resolved. Sorry if I delayed the merge unnecessarily.

**Comment by KaiSzuttor** [comment]
@kosovan as I said, that's not your fault. It's a bug of the mergebot.

**Pull request merged**
